contrib/smtp_proxy/filters/network/source/smtp_decoder.cc

```diff
@@ -0,0 +1,262 @@
+  if (crlf == -1) {
+    if (data.length() >= 512) {
+      return Decoder::Result::Bad;
+    } else {
+      return Decoder::Result::NeedMoreData;
+    }
+  }
+  ASSERT(crlf >= 0);
+  // empty line is invalid
+  if (static_cast<size_t>(crlf) == start) {
+    return Decoder::Result::Bad;
+  }
+
+  size_t len = crlf + CRLF.size() - start;
+  ASSERT(len > CRLF.size());
+  raw_line.reset(new char[len]);
+  data.copyOut(start, len, raw_line.get());
+  line = absl::string_view(raw_line.get(), len);
+
+  return Decoder::Result::ReadyForNext;
+}
+
+Decoder::Result DecoderImpl::DecodeCommand(Buffer::Instance& data, Command& result) {
+  std::unique_ptr<char[]> raw_line;
+  absl::string_view line;
+  Result res = GetLine(data, 0, raw_line, line, kMaxCommandLine);
+  if (res != Result::ReadyForNext) {
+    return res;
+  }
+  result.wire_len = line.size();
+
+  const bool trailing_crlf = absl::ConsumeSuffix(&line, CRLF);
+  ASSERT(trailing_crlf);
+
+  ssize_t space = line.find(' ');
+  if (space == -1) {
+    // no arguments e.g.
+    // NOOP\r\n
+    result.raw_verb = std::string(line);
+  } else {
+    // EHLO mail.example.com\r\n
+    result.raw_verb = std::string(line.substr(0, space));
+    result.rest = std::string(line.substr(space + 1));
+  }
+
+  static const struct {
+    Command::Verb verb;
+    absl::string_view raw_verb;
+  } kCommands[] = {
+      {Decoder::Command::HELO, "HELO"},
+      {Decoder::Command::EHLO, "EHLO"},
+      {Decoder::Command::STARTTLS, "STARTTLS"},
+  };
+
+  for (const auto& c : kCommands) {
+    if (absl::AsciiStrToUpper(result.raw_verb) != c.raw_verb) {
+      continue;
+    }
+    result.verb = c.verb;
+    break;
+  }
+
+  // We aren't too picky here since we're only going to look at the
+  // first ~2 commands (EHLO, STARTTLS) and the filter is going to
+  // hang up if the first isn't EHLO.
+  return Result::ReadyForNext;
+}
+
+Decoder::Result DecoderImpl::DecodeResponse(Buffer::Instance& data, Response& result) {
+  std::unique_ptr<char[]> raw_line;
+
+  absl::optional<int> code;
+  size_t line_off = 0;
+  std::string msg;
+  for (;;) {
+    absl::string_view line;
+    Result res = GetLine(data, line_off, raw_line, line, kMaxReplyLine);
+    if (res != Result::ReadyForNext) {
+      return res;
+    }
+    line_off += line.size();
+    if (line_off > kMaxReplyBytes) {
+      return Result::Bad;
+    }
+    // https://www.rfc-editor.org/rfc/rfc5321.html#section-4.2
+    //  Reply-line     = *( Reply-code "-" [ textstring ] CRLF )
+    //                 Reply-code [ SP textstring ] CRLF
+    //  Reply-code     = %x32-35 %x30-35 %x30-39
+    if (line.size() < 3) {
+      return Result::Bad;
+    }
+    for (int i = 0; i < 3; ++i) {
+      if (!absl::ascii_isdigit(line[i])) {
+        return Result::Bad;
+      }
+    }
+    absl::string_view code_str = line.substr(0, 3);
+    int this_code = 0;
+    const bool atoi_result = absl::SimpleAtoi(code_str, &this_code);
+    ASSERT(atoi_result);
```

Comment:
The same as above. What if someone sends malformed payload. Would the assertion fail?

Reply:
Internal invariant: the preceding ascii_isdigit() check restricts to a proper subset of what atoi accepts.

Reply:
OK. If you validated input and are sure that this will not fail, then it is OK.